File: day-11/soln.py
def solve(ipt):
    next_phase = ipt[:]
    for i, row in enumerate(ipt):
        for j, seat in enumerate(row):
            up = ipt[i-1][j] if i > 0 else None
            lup = ipt[i-1][j-1] if i > 0 and j > 0 else None
            rup = ipt[i-1][j+1] if i > 0 and j < len(row) - 1 else None
            right = ipt[i][j+1] if j < len(row) - 1 else None
            down = ipt[i+1][j] if i < len(ipt) - 1 else None
            ldown = ipt[i+1][j-1] if i < len(ipt) - 1 and j > 0 else None
            rdown = ipt[i+1][j+1] if i < len(ipt) - 1 and j < len(row) - 1 else None
            left = ipt[i][j-1] if j > 0 else None

            if seat == 'L' and all([x is None or x == '.' or x == 'L' for x in [up, right, down, left, ldown, lup, rdown, rup]]):
                next_phase[i] = next_phase[i][:j] + '#' + (next_phase[i][j+1:] if j < len(row) - 1 else '')
            elif seat == '#' and sum([1 for x in [up, right, down, left, ldown, lup, rdown, rup] if x == '#']) >= 4:
                next_phase[i] = next_phase[i][:j] + 'L' + (next_phase[i][j+1:] if j < len(row) - 1 else '')

    if ipt == next_phase:
        total = 0
        for row in ipt:
            for seat in row:
                if seat == '#':
                    total += 1

        print(total)
        return
    
    solve(next_phase)


def see_none(ipt, i, j):
    i1 = i + 1
    while i1 < len(ipt):
        if ipt[i1][j] == '#':
            return False
        elif ipt[i1][j] == 'L':
            break
        i1 += 1

    i1 = i - 1
    while i1 >= 0:
        if ipt[i1][j] == '#':
            return False
        elif ipt[i1][j] == 'L':
            break
        i1 -= 1

    j1 = j - 1
    while j1 >= 0:
        if ipt[i][j1] == '#':
            return False
        elif ipt[i][j1] == 'L':
            break
        j1 -= 1

    j1 = j + 1
    while j1 < len(ipt[0]):
        if ipt[i][j1] == '#':
            return False
        elif ipt[i][j1] == 'L':
            break
        j1 += 1

    i1 = i + 1
    j1 = j + 1
    while i1 < len(ipt) and j1 < len(ipt[0]):
        if ipt[i1][j1] == '#':
            return False
        elif ipt[i1][j1] == 'L':
            break
        i1 += 1
        j1 += 1

    i1 = i - 1
    j1 = j - 1
    while i1 >= 0 and j1 >= 0:
        if ipt[i1][j1] == '#':
            return False
        elif ipt[i1][j1] == 'L':
            break
        i1 -= 1
        j1 -= 1

    i1 = i - 1
    j1 = j + 1
    while i1 >= 0 and j1 < len(ipt[0]):
        if ipt[i1][j1] == '#':
            return False
        elif ipt[i1][j1] == 'L':
            break
        i1 -= 1
        j1 += 1
    
    i1 = i + 1
    j1 = j - 1
    while i1 < len(ipt) and j1 >= 0:
        if ipt[i1][j1] == '#':
            return False
        elif ipt[i1][j1] == 'L':
            break
        i1 += 1
        j1 -= 1

    return True


def see_some(ipt, i, j):
    total = 0

    i1 = i + 1
    while i1 < len(ipt):
        if ipt[i1][j] == '#':
            total += 1
            break
        elif ipt[i1][j] == 'L':
            break
        i1 += 1

    i1 = i - 1
    while i1 >= 0:
        if ipt[i1][j] == '#':
            total += 1
            break
        elif ipt[i1][j] == 'L':
            break
        i1 -= 1

    j1 = j - 1
    while j1 >= 0:
        if ipt[i][j1] == '#':
            total += 1
            break
        elif ipt[i][j1] == 'L':
            break
        j1 -= 1

    j1 = j + 1
    while j1 < len(ipt[0]):
        if ipt[i][j1] == '#':
            total += 1
            break
        elif ipt[i][j1] == 'L':
            break
        j1 += 1

    i1 = i + 1
    j1 = j + 1
    while i1 < len(ipt) and j1 < len(ipt[0]):
        if ipt[i1][j1] == '#':
            total += 1
            break
        elif ipt[i1][j1] == 'L':
            break
        i1 += 1
        j1 += 1

    i1 = i - 1
    j1 = j - 1
    while i1 >= 0 and j1 >= 0:
        if ipt[i1][j1] == '#':
            total += 1
            break
        elif ipt[i1][j1] == 'L':
            break
        i1 -= 1
        j1 -= 1

    i1 = i - 1
    j1 = j + 1
    while i1 >= 0 and j1 < len(ipt[0]):
        if ipt[i1][j1] == '#':
            total += 1
            break
        elif ipt[i1][j1] == 'L':
            break
        i1 -= 1
        j1 += 1
    
    i1 = i + 1
    j1 = j - 1
    while i1 < len(ipt) and j1 >= 0:
        if ipt[i1][j1] == '#':
            total += 1
            break
        elif ipt[i1][j1] == 'L':
            break
        i1 += 1
        j1 -= 1

    return total


def solve2(ipt):
    while True:
        next_phase = ipt[:]
        for i, row in enumerate(ipt):
            for j, seat in enumerate(row):
                if seat == 'L' and see_none(ipt, i, j):
                    next_phase[i] = next_phase[i][:j] + '#' + (next_phase[i][j+1:] if j < len(row) - 1 else '')
                elif seat == '#' and see_some(ipt, i, j) >= 5:
                    next_phase[i] = next_phase[i][:j] + 'L' + (next_phase[i][j+1:] if j < len(row) - 1 else '')

        if ipt == next_phase:
            total = 0
            for row in ipt:
                for seat in row:
                    if seat == '#':
                        total += 1

            print(total)
            return
        
        ipt = next_phase
    


if __name__ == '__main__':
    for f, t in [('test_input.txt', True), ('input.txt', False)]:
        ipt = []
        
        with open(f) as fp:
            ipt = [l.rstrip('\n') for l in fp]

        if t:
            print('TEST')
        else:
            print('RESULT')

        solve(ipt)
        solve2(ipt)
        print('')

File: day-11/soln.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strings"
)

type checker func(ipt [][]string, x int, y int) bool
type mover func(x int, y int) (int, int)

func checkAround(ipt [][]string, x int, y int) bool {
	for x1 := x-1; x1 < x+2; x1++ {
		if x1 < 0 || x1 >= len(ipt) {
			continue
		}

		for y1 := y-1; y1 < y+2; y1++ {
			if y1 < 0 || y1 >= len(ipt[0]) {
				continue
			}

			if x1 == x && y1 == y {
				continue
			}

			if ipt[x1][y1] == "#" {
				return false
			}
		}
	}

	return true
}

func countAround(ipt [][]string, x int, y int) (total int) {
	for x1 := x-1; x1 < x+2; x1++ {
		if x1 < 0 || x1 >= len(ipt) {
			continue
		}

		for y1 := y-1; y1 < y+2; y1++ {
			if y1 < 0 || y1 >= len(ipt[0]) {
				continue
			}

			if x1 == x && y1 == y {
				continue
			}

			if ipt[x1][y1] == "#" {
				total++
			}
		}
	}

	return
}

func copyInput(ipt [][]string) [][]string {
	dest := make([][]string, len(ipt))

	for i, row := range ipt {
		dest[i] = make([]string, len(row))
		for j, seat := range row {
			dest[i][j] = seat
		}
	}

	return dest
}

func up(x int, y int) (int, int) { return x - 1, y }
func right(x int, y int) (int, int) { return x, y + 1 }
func down(x int, y int) (int, int) { return x + 1, y }
func left(x int, y int) (int, int) { return x, y - 1 }

var directions = []mover{
	up, right, down, left,
	func (x int, y int) (int, int) { return up(right(x, y)) },
	func (x int, y int) (int, int) { return up(left(x, y)) },
	func (x int, y int) (int, int) { return down(right(x, y)) },
	func (x int, y int) (int, int) { return down(left(x, y)) },
}

type pair struct {
	x int
	y int
}

func moveDir(ipt [][]string, direction mover, x int, y int) (int, int, error) {
	for {
		x, y = direction(x, y)

		if x < 0 || x >= len(ipt) || y < 0 || y >= len(ipt[0]) {
			return -1, -1, errors.New("too far")
		}

		return x, y, nil
	}
}

func countSight(ipt [][]string, x int, y int) (total int) {
	for _, direction := range directions {
		for x1, y1, err := moveDir(ipt, direction, x, y);
			err == nil;
			x1, y1, err = moveDir(ipt, direction, x1, y1) {

			if ipt[x1][y1] == "#" {
				total++
				break
			} else if ipt[x1][y1] == "L" {
				break
			}
		}
	}
	
	return
}

func checkSight(ipt [][]string, x int, y int) bool {
	for _, direction := range directions {
		for x1, y1, err := moveDir(ipt, direction, x, y);
			err == nil;
			x1, y1, err = moveDir(ipt, direction, x1, y1) {
			if ipt[x1][y1] == "#" {
				return false
			} else if ipt[x1][y1] == "L" {
				break
			}
		}
	}	

	return true
}

func solve(ipt [][]string, shouldFill checker, shouldEmpty checker) {
	for {
		nextPhase := copyInput(ipt)

		for x, row := range ipt {
			for y, seat := range row {
				if seat == "L" && shouldFill(ipt, x, y) {
					nextPhase[x][y] = "#"
				} else if seat == "#" && shouldEmpty(ipt, x, y) {
					nextPhase[x][y] = "L"
				}
			}
		}

		changed := false
		for i, row1 := range ipt {
			for j, seat1 := range row1 {
				if seat1 != nextPhase[i][j] {
					changed = true
					break
				}
			}
			if changed {
				break
			}
		}

		if !changed {
			total := 0
			for _, row := range ipt {
				for _, seat := range row {
					if seat == "#" {
						total++
					}
				}
			}
			fmt.Println(total)
			return
		}

		ipt = nextPhase
	}
}

func main() {
	var ipt [][]string
	test, _ := os.Open("test_input.txt")
	defer test.Close()

	scanner := bufio.NewScanner(test)
	for scanner.Scan() {
		ipt = append(ipt, strings.Split(scanner.Text(), ""))
	}

	var part1CountChecker checker = func(ipt [][]string, x int, y int) bool {
		return countAround(ipt, x, y) >= 4
	}

	var part2CountChecker checker = func(ipt [][]string, x int, y int) bool {
		return countSight(ipt, x, y) >= 5
	}

	fmt.Println("TEST")
	solve(ipt, checkAround, part1CountChecker)
	solve(ipt, checkSight, part2CountChecker)
	fmt.Println("")

	ipt = nil
	real, _ := os.Open("input.txt")
	defer real.Close()

	scanner = bufio.NewScanner(real)
	for scanner.Scan() {
		ipt = append(ipt, strings.Split(scanner.Text(), ""))
	}

	fmt.Println("RESULT")
	solve(ipt, checkAround, part1CountChecker)
	solve(ipt, checkSight, part2CountChecker)
	fmt.Println("")
}